gst/nnstreamer/tensor_rate/gsttensorrate.c

```diff
@@ -0,0 +1,927 @@
+/* GObject vmethod implementations */
+static void gst_tensor_rate_set_property (GObject * object, guint prop_id,
+    const GValue * value, GParamSpec * pspec);
+static void gst_tensor_rate_get_property (GObject * object, guint prop_id,
+    GValue * value, GParamSpec * pspec);
+static void gst_tensor_rate_finalize (GObject * object);
+
+/* GstBaseTransform vmethod implementations */
+static GstFlowReturn gst_tensor_rate_transform_ip (GstBaseTransform * trans,
+    GstBuffer * buffer);
+static GstCaps *gst_tensor_rate_transform_caps (GstBaseTransform * trans,
+    GstPadDirection direction, GstCaps * caps, GstCaps * _rate);
+static GstCaps *gst_tensor_rate_fixate_caps (GstBaseTransform * trans,
+    GstPadDirection direction, GstCaps * caps, GstCaps * othercaps);
+static gboolean gst_tensor_rate_set_caps (GstBaseTransform * trans,
+    GstCaps * incaps, GstCaps * outcaps);
+static void gst_tensor_rate_swap_prev (GstTensorRate * self,
+    GstBuffer * buffer, gint64 time);
+static GstFlowReturn gst_tensor_rate_flush_prev (GstTensorRate * self,
+    gboolean duplicate, GstClockTime next_intime);
+
+static void gst_tensor_rate_notify_drop (GstTensorRate * self);
+static void gst_tensor_rate_notify_duplicate (GstTensorRate * self);
+
+static gboolean gst_tensor_rate_start (GstBaseTransform * trans);
+static gboolean gst_tensor_rate_stop (GstBaseTransform * trans);
+static gboolean gst_tensor_rate_sink_event (GstBaseTransform * trans,
+    GstEvent * event);
+
+static void gst_tensor_rate_install_properties (GObjectClass * gobject_class);
+
+/**
+ * @brief initialize the tensor_rate's class (GST Standard)
+ */
+static void
+gst_tensor_rate_class_init (GstTensorRateClass * klass)
+{
+  GObjectClass *gobject_class;
+  GstElementClass *gstelement_class;
+  GstBaseTransformClass *trans_class;
+
+  GST_DEBUG_CATEGORY_INIT (gst_tensor_rate_debug, "tensor_rate", 0,
+      "Tensor Rate to control streams based on tensor(s) values");
+
+  trans_class = (GstBaseTransformClass *) klass;
+  gstelement_class = (GstElementClass *) trans_class;
+  gobject_class = (GObjectClass *) gstelement_class;
+
+  gobject_class->set_property = gst_tensor_rate_set_property;
+  gobject_class->get_property = gst_tensor_rate_get_property;
+  gobject_class->finalize = gst_tensor_rate_finalize;
+
+  gst_tensor_rate_install_properties (gobject_class);
+
+  gst_element_class_set_details_simple (gstelement_class,
+      "Tensor_Rate",
+      "Tensor/Framerate",
+      "Adjusts a framerate of incoming tensors",
+      "Dongju Chae <dongju.chae@samsung.com>");
+
+  gst_element_class_add_pad_template (gstelement_class,
+      gst_static_pad_template_get (&src_factory));
+  gst_element_class_add_pad_template (gstelement_class,
+      gst_static_pad_template_get (&sink_factory));
+
+  trans_class->passthrough_on_same_caps = TRUE;
+  trans_class->transform_ip_on_passthrough = FALSE;
+
+  /* Processing units */
+  trans_class->transform_ip = GST_DEBUG_FUNCPTR (gst_tensor_rate_transform_ip);
```

Comment:
You use transform_ip only if you touch the contents. Refer to https://gstreamer.freedesktop.org/documentation/base/gstbasetransform.html?gi-language=c#modifications-only-to-the-capsmetadata-of-a-buffer

Reply:
IMO, when a framerate is changed, this element will think caps are also different and run on non-passthrough mode.
In this case, `transform` will make a unnecessary buffer copy. I think that's why existing rate-control elements (e.g., [videorate](https://gitlab.freedesktop.org/gstreamer/gst-plugins-base/blob/8ecc3b97307e3a47dc19a024de117cac68ff7d7b/gst/videorate/gstvideorate.c#L1286), [videomaxrate](https://code.cablelabs.com/DLNA-RUI/gst-plugins-bad/-/blob/de138dbc57c5928a25e10d440ff2bb0ac6ae4b4d/gst/videomaxrate/videomaxrate.c#L275)) use `transform-ip`.